### ISSUE TITLE
Federation error with "undrafted" articles

### ISSUE THREAD

**Issue by bannisdale** [Plume-org/Plume#221]
I was wondering why a certain post didn't show up in my Mastodon timeline.

I ran a few tests and realized: It's because it was a draft before. So from my tests, there are two scenarios:

1. Post an article. It gets federated (shows up in my Mastodon feed)
2. Save an article as draft. Then edit it and publish it ("undraft" it). It won't show up.

I assume it's because the post is initially federated (without visibility) and technically edited later on, but Mastodon doesn't support editing.

Workaround so far is writing the draft until it's final, then copy the text, create a new post with it and then delete the old draft. Obviously, that's a bit lackluster. :)

**Comment by kvibber** [comment]
Additionally, I found that changing the title between the first draft and the published draft updates the URL of the blog post, but doesn't update the link in the federated post that gets into Mastodon if you search for the URL to manually pull it in. So you can make the post title and link visible, but the link is broken, because it's pointing at the old URL.

I followed @bannisdale's workaround and deleted/re-created the post to get around it.

**Comment by BaptisteGelez** [comment]
@kvibber there is another issue for your problem there: #207 :slightly_smiling_face:

**Comment by kvibber** [comment]
@BaptisteGelez Thanks, I didn't see that one!

Looking at the interaction between the two bugs, it seems it would make sense to update the URL when changing a draft's title, but not when changing a published post's title.

**Comment by trwnh** [comment]
Ideally Mastodon would support Update activities, but there is another solution: federation permalink can be ID-based instead of title-based, and a "pretty url" can be formed from a custom slug containing the title or whatever the user enters.

So, for example, I might post an article on Plume that federates out as `https://plume.example/~/trwnh/6743826481264` and the title is `test post`. The default pretty-url slug would be `test-post` (but I could edit it manually to `test` if I wanted to, or `2018/10/11/1`, or `2018/testpost`, or whatever I wanted) and then the URL would become `https://plume.example/~/trwnh/test-post` (or whatever i set it to)

When federating that out, you would use the `id` as that permalink, and `url` as the slugified link. So for example:
```
"id": "https://plume.example/~/trwnh/6743826481264",
"url": "https://plume.example/~/trwnh/test-post",
```

And the `content` of the `Create` activity would include the `id` instead of the `url`. So something like `<p>$title - $id</p>` at the simplest, and then it would show up in Mastodon like this:

```
trwnh (@trwnh@plume.example) posted 23s ago:
test post - https://plume.example/~/trwnh/6743826481264
```

or `trwnh posted a new article:<br>$title<br>$id`, and then it would show up in Mastodon like this:

```
trwnh (@trwnh@plume.example) posted 23s ago:
trwnh posted a new article:
test post
https://plume.example/~/trwnh/6743826481264
```

The important thing is that the URL that gets federated out cannot change, so it's a bad idea to use a value that can be changed.

**Comment by fdb-hiroshima** [comment]
This is now fixed, #207 is too, however @trwnh's suggestion might be a better way of handling post title update. Shall we open a new issue discussing if this should be done?

**Comment by trwnh** [comment]
@fdb-hiroshima I can do that, if you wish.
